README.md
Mirage
======

[![Build Status](https://travis-ci.org/SpectoLabs/mirage.png?branch=master)](https://travis-ci.org/SpectoLabs/mirage)
[![Documentation Status](https://readthedocs.org/projects/mirage/badge/?version=latest)](http://mirage.readthedocs.org/en/latest/?badge=latest)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/SpectoLabs/mirage?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=body_badge)

Mirage enables development and automated testing by allowing you to virtualize your application dependencies.

## Quickstart

For the really impatient ones - use our Tutum stack configuration (includes load balanced HAProxy environment with MongoDB and Redis dependencies): 

[![Deploy to Tutum](https://s.tutum.co/deploy-to-tutum.svg)](https://dashboard.tutum.co/stack/deploy/)

For less impatient, [instructions are here.](https://github.com/SpectoLabs/mirage/wiki/Installation)

[Try the demo project](https://github.com/SpectoLabs/mirage_demo) to see how Mirage works with applications and third-party APIs.

## Documentation

[New documentation](https://github.com/SpectoLabs/mirage/wiki)

[Old documentation](<http://mirage.readthedocs.org/en/latest/?badge=latest/>)

## Authors

Mirage is actively maintained by [SpectoLabs](http://specto.io).

For support, [contact us.](mailto:contact@specto.io)

## License

Mirage is made available under GPLv3. See LICENSE for more details.


CHANGES.txt
Changelog
=========

Added:

Changed:

Fixed:

Deprecated:

Removed:

0.8.2 (2015-11-05)

Added:
- Displaying applied delays in event log (tracker) page if there was one. Users can now filter based on added delays
  by adding "d:200" to display all 200 ms delays, operators like >, <, >=, <= are also available.

0.8.1 (2015-11-04)

Added:
- Database authentication. Now you can specify MongoDB user (mongo.user) and password (mongo.password) in the .ini file.
- Now looking for environment variables for database connection setup:
  + export MONGO_URI=mongodb://<dbuser>:<dbpassword>@mongo_host:mongo_port/database
  + export MONGO_DB=stubodb
- Cache (Redis) connection info can now be supplied through environment variables:
  + export REDIS_ADDRESS=new_hostname_or_ip
  + export REDIS_PORT=6379
  + export REDIS_PASSWORD=very_secret
  More information can be found here: https://github.com/SpectoLabs/mirage/wiki/Configuration

Changed:

Fixed:
- Track all hosts not showing information for all hostnames due to bug when loading initial cookie.


0.8.0 (2015-10-27)

Added:
- Documentation for Tracker API and how to create useful queries. Also, created documentation for Tracker record details.

Changed:
- Disabled certificate verification when importing scenarios (executing commands)
- Changed all REST API (v2) URL paths: removed "/stubo/" from URL since it doesn't hold any information and there is no
  point in providing application former name inside url.
- Updated documentation, tests to accommodate URL changes.

0.7.2 (2015-10-26)

Added:
- Direct upload API (addresses https://github.com/SpectoLabs/mirage/issues/10). Now users can upload exported .zip files
  directly into Mirage. There were also few modifications how Mirage parses .yaml configuration files such as
  now it only reads scenario and session names. All archived .json files will be treated as stubs and will be
  imported under the same scenario. This will make it easier to generate initial test data since you don't have
  to updated .yaml configuration file with each created stub file. Multiple PRs.

Changed:
- Changed how Mirage exports scenarios to ".yaml" files. Now it doesn't add some non compatible tags like {{values}} to yaml file since
  readers couldn't parse that information. Those lines were added to export files so it would be possible to change values
  during import but it is a lot easier to do that with standard "yaml" libraries.
- Delete scenario button is now disabled by default if scenario has active sessions in "playback" or "record" mode.
- Updates to documentation https://github.com/SpectoLabs/mirage/pull/62,  https://github.com/SpectoLabs/mirage/pull/61.

Fixed:
- Fixed bug when you could have deleted scenario which had active sessions. PR: https://github.com/SpectoLabs/mirage/pull/49

Removed:
- Documentation cleanup in https://github.com/SpectoLabs/mirage/pull/62,  https://github.com/SpectoLabs/mirage/pull/61. A lot
  of content added to project wiki https://github.com/SpectoLabs/mirage/wiki. Since it is a lot easier to maintain wiki than
  sphinx docs - only essential, tightly coupled information like API reference and few other things were left in traditional
  docs.

0.7.1 (2015-10-20):

Added:
- Added button to create scenario where users can also specify a session name and Mirage will start recording.
- "End all sessions" button now instead of being disabled for dormant sessions - switches either to "begin playback"
  or "start recording" based on current stub count.
- Added new API for getting information about found response (similar to get stub response) although instead of Mirage
  just using that information to create the original response - it now provides data about desired headers, body and
  status code. This enables proxy to store encoded data inside Mirage with required headers for decoding and then
  recreating original response without tampering with it. Original design was limited to plain text only.
- Users can now override session and scenario by supplying any hostname. So a user that is accessing Mirage through the
  hostname A - can also create scenario, start session and end it for host B. This was needed for running Mirage in container

stubo/model/db.py
"""  
    :copyright: (c) 2015 by OpenCredo.
    :license: GPLv3, see LICENSE for more details.
"""
from pymongo import MongoClient, DESCENDING, ASCENDING
import logging
from bson.objectid import ObjectId
from stubo.utils import asbool
from stubo.model.stub import Stub
import hashlib
import time
import motor
import os

default_env = {
    'port': 27017,
    'max_pool_size': 20,
    'tz_aware': True,
    'db': 'stubodb'
}


def coerce_mongo_param(k, v):
    if k in ('port', 'max_pool_size'):
        return int(v)
    elif k in ('tz_aware',):
        return asbool(v)
    return v


log = logging.getLogger(__name__)

mongo_client = None


def motor_driver(settings):
    """

    Returns asynchronous Motor client. If user and password provided in config file - returns authenticated connection
    :param settings:
    :return:
    """
    # checking for environment variables
    mongo_uri = os.getenv("MONGO_URI")
    mongo_db = os.getenv("MONGO_DB")
    if mongo_uri and mongo_db:
        client = motor.MotorClient(mongo_uri)
        log.info("MongoDB environment variables found: %s!" % mongo_uri)
        return client[mongo_db]

    # environment variables not found, looking for details from configuration file
    user = settings.get('mongo.user', None)
    password = settings.get('mongo.password', None)
    if user and password:
        uri = "mongodb://{user}:{password}@{host}:{port}/{database_name}".format(
            user=user,
            password=password,
            host=settings['mongo.host'],
            port=settings['mongo.port'],
            database_name=settings['mongo.db']
        )
        client = motor.MotorClient(uri)

    else:
        client = motor.MotorClient(settings['mongo.host'], int(settings['mongo.port']))
    return client[settings['mongo.db']]


def get_mongo_client():
    return mongo_client


def get_connection(env=None):
    """

    Gets MongoDB connection. If user and password provided - authenticates (logs in)
    :param env: dictionary, example:
            {'host': 'ds045454.mongolab.com',
             'tz_aware': True,
             'max_pool_size': 10,
             'port': 45454}
    :return: MongoClient
    """
    # checking for environment variables
    mongo_uri = os.getenv("MONGO_URI")
    mongo_db = os.getenv("MONGO_DB")
    if mongo_uri and mongo_db:
        client = MongoClient(mongo_uri)
        log.info("MongoDB environment variables found: %s!" % mongo_uri)
        return client[mongo_db]

    # environment variables not found, looking for details from configuration file
    env = env or default_env
    _env = env.copy()
    dbname = _env.pop('db', None)

    # if auth details supplied - getting details
    user = password = None
    if 'user' in _env:
        user = _env.pop('user')
    if 'password' in _env:
        password = _env.pop('password')

    client = MongoClient(**_env)
    if dbname:
        log.debug('using db={0}'.format(dbname))
        client = getattr(client, dbname)

    # authenticating
    if user and password:
        # if fails - throws exception which will be handled in run_stubo.py
        client.authenticate(user, password)
        log.info("Login to MongoDB successful!")

    return client


class Scenario(object):
    def __init__(self, db=None):
        self.db = db or mongo_client
        assert self.db

    def get_stubs(self, name=None):
        if name:
            filter = {'scenario': name}
            return self.db.scenario_stub.find(filter)
        else:
            return self.db.scenario_stub.find()

    def get_pre_stubs(self, name=None):
        if name:
            query = {'scenario': name}
            return self.db.pre_scenario_stub.find(query)
        else:
            return self.db.scenario_pre_stub.find()

    def stub_count(self, name):
        return self.get_stubs(name).count()

    def get(self, name):
        return self.db.scenario.find_one({'name': name})

    def get_all(self, name=None):
        if name:
            cursor = self.db.scenario.find({'name': name})
        else:
            cursor = self.db.scenario.find()
        return cursor

    def insert(self, **kwargs):
        return self.db.scenario.insert(kwargs)

    def change_name(self, name, new_name):
        """
        Rename scenario and all stubs
        :param name: current scenario name
        :param new_name: new scenario name
        :return: statistics, how many stubs were changed
        """
        # updating scenario stub collection. You have to specify all parameters as booleans up to the one that you
        # actually want, in our case - the fourth parameter "multi" = True
        # update(spec, document[, upsert=False[,
        #                        manipulate=False[, safe=None[, multi=False[, check_keys=True[, **kwargs]]]]]])
        response = {
            'Old name': name,
            "New name": new_name
        }
        try:
            result = self.db.scenario_stub.update(
                {'scenario': name}, {'$set': {'scenario': new_name}}, False, False, None, True)
            try:
                response['Stubs changed'] = result['nModified']
            except KeyError:
                # older versions of mongodb returns 'n' instead of 'nModified'
                response['Stubs changed'] = result['n']
            except Exception as ex1:
                # this is probably KeyError, leaving Exception for debugging purposes
                log.debug("Could not get STUB nModified key, result returned: %s. Error: %s" % (result, ex1))
        except Exception as ex:
            log.debug("Could not update scenario stub, got error: %s " % ex)
            response['Stubs changed'] = 0

        # updating pre stubs
        try:
            result = self.db.scenario_pre_stub.update(
                {'scenario': name}, {'$set': {'scenario': new_name}}, False, False, None, True)
            try:
                response['Pre stubs changed'] = result['nModified']
            except KeyError:
                # older versions of mongodb returns 'n' instead of 'nModified'
                response['Pre stubs changed'] = result['n']
            except Exception as ex1:
                log.debug("Could not get PRE STUB nModified key, result returned: %s. Error: %s" % (result, ex1))
        except Exception as ex:
            log.debug("Could not update scenario pre stub, got error: %s" % ex)
            response['Pre stubs changed'] = 0

        try:
            # updating scenario itself
            result = self.db.scenario.update({'name': name}, {'name': new_name})
            try:
                response['Scenarios changed'] = result['nModified']
            except KeyError:
                # older versions of mongodb returns 'n' instead of 'nModified'
                response['Scenarios changed'] = result['n']
            except Exception as ex1:
                log.debug("Could not get SCENARIO nModified key, result returned: %s. Error: %s" % (result, ex1))
        except Exception as ex:
            log.debug("Could not update scenario, got error: %s" % ex)
            response['Scenarios changed'] = 0

        return response

    def recorded(self, name=None):
        """
        Calculates scenario recorded date. If name is not supplied - returns a dictionary with scenario name and
        recorded date:
        { 'scenario_1': '2015-05-07',
          'scenario_2': '2015-05-07'}
        If a name is supplied - returns recorded date string (since MongoDB does not support DateTimeField).
        :param name: optional parameter to get recorded date for specific scenario
        :return: <dict> - if name is not supplied, <string> with date - if scenario name supplied.
        """
        start_time = time.time()
        pipeline = [
            {'$group': {
                '_id': '$scenario',
                'recorded': {'$max': '$recorded'}}}]
        # use the pipe to calculate latest date
        try:
            result = self.db.command('aggregate', 'scenario_stub', pipeline=pipeline)['result']
        except KeyError as ex:
            log.error(ex)
            return None
        except Exception as ex:
            log.error("Got error when trying to use aggregation framework: %s" % ex)
            return None
        # using dict comprehension to form a new dict for fast access to elements
        result_dict = {x['_id']: x['recorded'] for x in result}

        # finish time
        finish_time = time.time()
        log.info("Recorded calculated in %s ms" % int((finish_time - start_time) * 1000))
        # if name is provided - return only single recorded date for specific scenario.
        if name:
            scenario_recorded = None
            try:
                scenario_recorded = result_dict[name]
            except KeyError:
                log.debug("Wrong scenario name supplied (%s)" % name)
            except Exception as ex:
                log.warn("Failed to get scenario recorded date for: %s, error: %s" % (name, ex))
            return scenario_recorded
        else:
            # returning full list (scenarios and sizes)
            return result_dict

    def size(self, name=None):
        """
        Calculates scenario sizes. If name is not supplied - returns a dictionary with scenario name and size
        (in bytes):
        { 'scenario_1': 5646145,
          'scenario_2': 12312312}
        If a name is supplied - returns a size Integer in bytes.
        :param name: optional parameter to get size of specific scenario
        :return: <dict> - if name is not supplied, <int> - if scenario name supplied.
        """
        start_time = time.time()
        pipeline = [{
            '$group': {
                '_id': '$scenario',
                'size': {'$sum': {'$divide': ['$space_used', 1024]}}
            }
        }]

        # use the pipe to calculate scenario sizes
        try:
            result = self.db.command('aggregate', 'scenario_stub', pipeline=pipeline)['result']
        except KeyError as ex:
            log.error(ex)
            return None
        except Exception as ex:
            log.error("Got error when trying to use aggregation framework: %s" % ex)
            return None
        # using dict comprehension to form a new dict for fast access to elements
        result_dict = {x['_id']: x['size'] for x in result}

        # finish time
        finish_time = time.time()
        log.info("Sizes calculated in %s ms" % int((finish_time - start_time) * 1000))
        # if name is provided - return only single size for specific scenario.
        if name:
            scenario_size = None
            try:
                scenario_size = result_dict[name]
            except KeyError:
                log.debug("Wrong scenario name supplied (%s)" % name)
            except Exception as ex:
                log.warn("Failed to get scenario size for: %s, error: %s" % (name, ex))
            return scenario_size
        else:
            # returning full list (scenarios and sizes)
            return result_dict

    def stub_counts(self):
        """
        Calculates stub counts:
        { 'scenario_1': 100,
          'scenario_2': 20}

        Remember, that if the scenario doesn't have any stubs - it will not be in this list since it is not accessing
        scenario collection to add scenarios with 0 stubs.
        :return: <dict>
        """
        start_time = time.time()
        pipeline = [{'$group': {
            '_id': '$scenario',
            'count': {'$sum': 1}
        }
        }]

        # use the pipe to calculate scenario stub counts
        try:
            result = self.db.command('aggregate', 'scenario_stub', pipeline=pipeline)['result']
        except KeyError as ex:
            log.error(ex)
            return None
        except Exception as ex:
            log.error("Got error when trying to use aggregation framework: %s" % ex)
            return None
        # using dict comprehension to form a new dict for fast access to elements
        result_dict = {x['_id']: x['count'] for x in result}

        # finish time
        finish_time = time.time()
        log.info("Stub counts calculated in %s ms" % int((finish_time - start_time) * 1000))

        return result_dict

    @staticmethod
    def _create_hash(matchers):
        """
        Creates a hash out of matchers list.
        :param matchers: <list> matchers
        :return: matchers md5 hash
        """
        if matchers is not None:
            return hashlib.md5(u"".join(unicode(matchers))).hexdigest()
        elif matchers is None:
            return None

    def get_matched_stub(self, name, matchers_hash):
        """
        Gets matched stub for specific scenario. Relies on indexed "matchers" key in scenario_stub object
        :param name: <string> scenario name
        :param matchers_hash: <string> matcher hash
        :return: matched stub document or None if stub not found
        """
        if name:
            pattern = {'scenario': name,
                       'matchers_hash': matchers_hash}

            return self.db.scenario_stub.find_one(pattern)

    def insert_stub(self, doc, stateful):
        """
        Insert stub into DB. Performs a check whether this stub already exists in database or not.  If it exists
        and stateful is True - new response is appended to the response list, else - reports that duplicate stub
        found and it will not be inserted.

        :param doc: Stub class with Stub that will be inserted
        :param stateful: <boolean> specify whether stub insertion should be stateful or not
        :return: <string> message with insertion status:
           ignored - if not stateful and stub was already present
           updated - if stateful and stub was already present
           created - if stub was not present in database
        """
        # getting initial values - stub matchers, scenario name
        matchers = doc['stub'].contains_matchers()
        scenario = doc['scenario']

        matchers_hash = self._create_hash(matchers)
        # check if we have matchers - should be None for REST calls
        if matchers is not None:
            # additional helper value for indexing
            doc['matchers_hash'] = matchers_hash
            matched_stub = self.get_matched_stub(name=scenario, matchers_hash=matchers_hash)
            # checking if stub already exists
            if matched_stub:
                # creating stub object from found document
                the_stub = Stub(matched_stub['stub'], scenario)
                if not stateful and doc['stub'].response_body() == the_stub.response_body():
                    msg = 'duplicate stub found, not inserting.'
                    log.warn(msg)
                    result = {'status': 'ignored',
                              'msg': msg,
                              'key': str(matched_stub['_id'])}
                    return result
                # since stateful is true - updating stub body by extending the list
                log.debug('In scenario: {0} found exact match for matchers:'
                          ' {1}. Perform stateful update of stub.'.format(scenario, matchers))
                response = the_stub.response_body()
                response.extend(doc['stub'].response_body())
                the_stub.set_response_body(response)
                # updating Stub body and size, writing to database
                self.db.scenario_stub.update(
                    {'_id': matched_stub['_id']},
                    {'$set': {'stub': the_stub.payload,
                              'space_used': len(unicode(the_stub.payload))}})
                result = {'status': 'updated',
                          'msg': 'Updated with stateful response',
                          'key': str(matched_stub['_id'])}
                return result

        # inserting stub into DB
        status = self.db.scenario_stub.insert(self.get_stub_document(doc))

        result = {'status': 'created',
                  'msg': 'Inserted scenario_stub',
                  'key': str(status)}
        return result

    @staticmethod
    def get_stub_document(doc):
        """

        prepares stub document for insertion into database
        :param doc:
        :return:
        """
        doc['stub'] = doc['stub'].payload

        # additional helper for aggregation framework
        try:
            doc['recorded'] = doc['stub']['recorded']
        except KeyError:
            # during tests "recorded" value is not supplied
            pass
        # calculating stub size
        doc['space_used'] = len(unicode(doc['stub']))

        return doc

    def insert_pre_stub(self, scenario, stub):
        status = self.db.pre_scenario_stub.insert(dict(scenario=scenario,
                                                       stub=stub.payload))
        return 'inserted pre_scenario_stub: {0}'.format(status)

    def remove_all(self, name):
        self.db.scenario.remove({'name': name})
        self.db.scenario_stub.remove({'scenario': name})
        self.db.pre_scenario_stub.remove({'scenario': name})

    def remove_all_older_than(self, name, recorded):
        # recorded = yyyy-mm-dd
        self.db.scenario_stub.remove({
            'scenario': name,
            'recorded': {"$lt": recorded}
        })
        self.db.pre_scenario_stub.remove({
            'scenario': name,
            'recorded': {"$lt": recorded}
        })
        if not self.stub_count(name):
            self.db.scenario.remove({'name': name})


class Tracker(object):
    def __init__(self, db=None):
        self.db = db or mongo_client

    def insert(self, track, write_concern=0):
        # w=0 disables write ack 
        """
        Insert tracker doc into MongoDB and creates indexes for faster search.
        :param track: tracker object
        :param write_concern: 1 or 0, check mongo docs for more info
        :return:
        """
        forced_log_id = track.get('forced_log_id')
        if forced_log_id:
            track['_id'] = int(forced_log_id)

        return self.db.tracker.insert(track, w=write_concern)

    def _create_index(self, key=None, direction=DESCENDING):
        """
        Creates index for specific key, fails silently if index creation was unsuccessful. Key examples:
        "host" , "scenario", "scenario", "request_params.session"
        :param key: <string>
        :param direction: ASCENDING or DESCENDING (from pymongo)
        """
        if key:
            try:
                self.db.tracker.create_index(key, direction)
            except Exception as ex:
                log.debug("Could not create index (tracker collection) for key %s, got error: %s" % (key, ex))

    def find_tracker_data(self, tracker_filter, skip, limit):
        projection = {'start_time': 1, 'function': 1, 'return_code': 1, 'scenario': 1,
                   'stubo_response': 1, 'duration_ms': 1, 'request_params.session': 1,
                   'delay': 1}
        if skip < 0:
            skip = 0
        # sorted on start_time descending    
        return self.db.tracker.find(tracker_filter, projection).sort('start_time',
                                                                     -1).limit(limit).skip(skip)

    def item_count(self, query=None):
        """

        Total item counter for tracker collection.
        :param query: query dict, example : { return_code: { $ne: 200 } }
        :return:
        """
        if query is not None:
            return self.db.tracker.find(query).count()
        else:
            return self.db.tracker.count()

    def find_tracker_data_full(self, _id):
        return self.db.tracker.find_one({'_id': ObjectId(_id)})

    def session_last_used(self, scenario, session, mode):
        """
        Return the date this session was last used using the
            last put/stub time (for record) or last get/response time otherwise.
        """
        if mode == 'record':
            function = 'put/stub'
        else:
            function = 'get/response'
        host, scenario_name = scenario.split(':')
        return self.db.tracker.find_one({
            'host': host,
            'scenario': scenario_name,
            'request_params.session': session,
            'function': function}, sort=[("start_time", DESCENDING)])

    def get_last_playback(self, scenario, session, start_time):
        start = self.db.tracker.find_one({
            'scenario': scenario,
            'request_params.session': session,
            'request_params.mode': 'playback',
            'function': 'begin/session',
            'start_time': {"$lt": start_time}
        }, {'start_time': 1}, sort=[("start_time", DESCENDING)])
        end = self.db.tracker.find_one({
            'scenario': scenario,
            'request_params.session': session,
            'function': 'end/session',
            'start_time': {"$gt": start_time}
        }, {'start_time': 1}, sort=[("start_time", DESCENDING)])
        if not (start or end):
            return []

        project = {'start_time': 1, 'return_code': 1, 'stubo_response': 1,
                   'response_headers': 1, 'request_headers': 1, 'duration_ms': 1,
                   'request_params': 1, 'request_text': 1, 'delay': 1}
        query = {
            'scenario': scenario,
            'request_params.session': session,
            'function': 'get/response',
            'start_time': {"$gt": start['start_time'],
                           "$lt": end['start_time']}
        }
        return self.db.tracker.find(query, project).sort("start_time",
                                                         ASCENDING)

    def get_last_recording(self, scenario, session, end):
        # find the last begin/session?mode=record from the last put/stub time 
        start = self.db.tracker.find_one({
            'scenario': scenario,
            'request_params.session': session,
            'request_params.mode': 'record',
            'function': 'begin/session',
            'start_time': {"$lt": end}
        }, {'start_time': 1}, sort=[("start_time", DESCENDING)])
        if not start:
            return []

        project = {'start_time': 1, 'return_code': 1, 'stubo_response': 1,
                   'response_headers': 1, 'request_headers': 1, 'duration_ms': 1,
                   'request_params': 1, 'request_text': 1, 'delay': 1}
        # get all the put/stubs > last begin/session?mode=record and <= last put/stub   
        query = {
            'scenario': scenario,
            'request_params.session': session,
            'function': 'put/stub',
            'start_time': {"$gt": start['start_time'],
                           "$lte": end}
        }
        log.debug('tracker.find: {0}'.format(query))
        return self.db.tracker.find(query, project).sort("start_time",
                                                         ASCENDING)


def session_last_used(scenario, session_name, mode):
    tracker = Tracker()
    return tracker.session_last_used(scenario, session_name, mode)
